core/src/ast/data.rs

```diff
@@ -219,52 +254,94 @@ impl<T> Fields<T> {
     }
 }
 
-impl<F: FromField> Fields<F> {
-    pub fn try_from(fields: &syn::Fields) -> Result<Self> {
-        let (items, errors) = match *fields {
-            syn::Fields::Named(ref fields) => {
-                let mut items = Vec::with_capacity(fields.named.len());
-                let mut errors = Vec::new();
-
-                for field in &fields.named {
-                    match FromField::from_field(field) {
-                        Ok(val) => items.push(val),
-                        Err(err) => errors.push(if let Some(ref ident) = field.ident {
-                            err.at(ident)
-                        } else {
-                            err
-                        }),
+impl<T: ToTokens> ToTokens for Fields<T> {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        let fields = &self.fields;
+        // An unknown Span should be `Span::call_site()`;
+        // https://docs.rs/syn/1.0.12/syn/spanned/trait.Spanned.html#tymethod.span
+        let span = self.span.unwrap_or_else(|| Span::call_site());
+
+        match self.style {
+            Style::Struct => {
+                let trailing_comma = {
+                    if fields.is_empty() {
+                        quote!()
+                    } else {
+                        quote!(,)
                     }
-                }
+                };
 
-                (items, errors)
+                tokens.extend(quote_spanned![span => { #(#fields),* #trailing_comma }]);
             }
-            syn::Fields::Unnamed(ref fields) => {
-                let mut items = Vec::with_capacity(fields.unnamed.len());
-                let mut errors = Vec::new();
+            Style::Tuple => {
+                tokens.extend(quote_spanned![span => ( #(#fields),* )]);
+            }
+            Style::Unit => {}
+        }
+    }
+}
 
-                for field in &fields.unnamed {
-                    match FromField::from_field(field) {
-                        Ok(val) => items.push(val),
-                        Err(err) => errors.push(if let Some(ref ident) = field.ident {
-                            err.at(ident)
-                        } else {
-                            err
-                        }),
+impl<T: PartialEq> PartialEq for Fields<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.style == other.style && self.fields == other.fields
+    }
+}
+
+impl<T: Eq> Eq for Fields<T> {}
+
+impl<F: FromField> Fields<F> {
```

Comment:
Minor nit: I'd prefer to keep the inherent impl block above the trait impls. That should also reduce the diff size.